File: code/SEER_1_0/python/runSEER.py
import numpy as np
from scipy.sparse import lil_matrix
import scipy.io as sio

#Implementation of Sparse Exemplar Ensemble Representations (SEER) 

#Peer Neubert and Stefan Schubert (2022), "SEER: Unsupervised and 
#sample-efficient environment specialization of image descriptors", 
#Proceeding of Robotics: Science And Systems, New York, USA

#This is a batch version that computes output descriptors for a batch of
#input descriptors Y.

#M ... sparse exemplar memory, each COLUMN is an exemplar, can be empty. It
#          is important to allocate the appropriate amount of memory. If an
#          empty matrix is provided, then the memory is allocated within this
#          function.
#Y ... input descriptor, each row is a descriptor. Should be a distributed 
#          representation.
#updatedM_flag ... toggles whether new rows/exemplars are added to M
#d_M ... number of non-zero elements in the exemplars in M, default: 200
#k ... minimum number of exemplars per database descriptor, default: 50
#lambda ... factor on k for non-zero elements in the output descriptor, default: 2

#M ... see input, will be updated if updatedM_flag is set
#DS ... sparse output descriptor

#peer.neubert@etit.tu-chemnitz.de, 2022

def run_seer(M, Y, update_M_flag, d_M=200, k=50, lambda_val=2):
    
    # Allocate memory for output descriptor
    DS = lil_matrix((Y.shape[0], M.shape[1] if M is not None else 0), dtype=np.float32)
    
    # Allocate memory for M if it is empty AND there has not already been allocated some memory
    if M is None or M.shape[1] == 0:
        M = lil_matrix((Y.shape[1], 1), dtype=np.float32)
    
    print(M.shape)
    # L2 normalize each input descriptor
    Y = Y / np.linalg.norm(Y, axis=1, keepdims=True)
    
    # Process each descriptor
    for i in range(Y.shape[0]):
        if i % 100 == 0:
            print(f'  running image {i+1} of {Y.shape[0]} ({M.shape[1]} exemplars) ')
        
        # Compute similarity
        if M.shape[1] > 0:
            S = Y[i, :] @ M.toarray()
        else:
            S = np.array([])
        
        # Sparsify
        if S.size > 0:
            knn_sim = np.partition(S, -lambda_val*k)[-lambda_val*k:]
            knn_idx = np.argpartition(S, -lambda_val*k)[-lambda_val*k:]
        else:
            knn_sim = np.array([])
            knn_idx = np.array([])
        
        if update_M_flag:
            # Apply threshold
            expected_similarity = d_M / Y.shape[1]
            thresh_idx = knn_sim >= expected_similarity
            knn_idx_above_thresh = knn_idx[thresh_idx]
            print("reached line 67")
        
            # Potentially create new patterns in M        
            if knn_idx_above_thresh.size < k:
                n_new_patterns = k - knn_idx_above_thresh.size
                PI, PV = create_exemplars(Y[i, :], d_M, n_new_patterns)
                print("PI shape:")
                print(PI.shape)
                new_pattern_idx = np.arange(M.shape[1], M.shape[1] + PI.shape[0])
                new_pattern_activity = np.sum(PV * PV, axis=1)
                print("reached line 75")
                for p_idx in range(PI.shape[0]):
                    M = np.hstack((M, lil_matrix((Y.shape[1], 1), dtype=np.float32)))
                    M[PI[p_idx], -1] = PV[p_idx]
            else:
                new_pattern_idx = np.array([])
                new_pattern_activity = np.array([])
                n_new_patterns = 0
        else:
            new_pattern_idx = np.array([])
            new_pattern_activity = np.array([])
            n_new_patterns = 0
        
        # Output descriptor is kNN results above threshold and new patterns
        n_remaining_kNN = min(knn_idx.size, k * lambda_val - n_new_patterns)
        descr_idx = np.hstack((knn_idx[:n_remaining_kNN], new_pattern_idx))
        descr_activity = np.hstack((knn_sim[:n_remaining_kNN], new_pattern_activity))
                
        # Create output descriptor
        DS[i, descr_idx] = descr_activity
    
    return M, DS

def create_exemplars(input_Y, n_dim_samples, n_patterns):
    """
    Sample new exemplars

    input_Y ... continuous input descriptor
    PI ... index matrix, each row is vector of size nDimSamples
    PV ... the values from inputY that correspond to these indexes 
    """
    PI = np.zeros((n_patterns, n_dim_samples), dtype=int)
    for i in range(n_patterns):
        if n_dim_samples == input_Y.shape[0]:
            PI[i, :] = np.arange(input_Y.shape[0])
        else:
            PI[i, :] = np.random.choice(input_Y.shape[0], n_dim_samples, replace=False, p=np.abs(input_Y)/np.sum(np.abs(input_Y)))
    
    PV = input_Y[PI]
    return PI, PV

if __name__=="__main__":
    DB = sio.loadmat('../data/HDC_DELF_GardensPointWalking_day_left.mat')['Y']
    Q = sio.loadmat('../data/HDC_DELF_GardensPointWalking_night_right.mat')['Y']
    gt = sio.loadmat('../data/groundTruth_GPW_DL_NR.mat')['GT']
    M = None
    M, DS = run_seer(M, DB, True, 200, 50, 2)
    _, Q_SEER = run_seer(M, Q, False, 200, 50, 2)
    
# Beispielnutzung:
# import scipy.io as sio
# DB = sio.loadmat('data/HDC_DELF_GardensPointWalking_day_left.mat')['Y']
# Q = sio.loadmat('data/HDC_DELF_GardensPointWalking_night_right.mat')['Y']
# gt = sio.loadmat('data/groundTruth_GPW_DL_NR.mat')['GT']
# M = None
# M, DS = run_seer(M, DB, True, 200, 50, 2)
# _, Q_SEER = run_seer(M, Q, False, 200, 50, 2)

File: code/SEER_1_0/MATLAB/runSEER.m
% Implementation of Sparse Exemplar Ensemble Representations (SEER) 
%
%    Peer Neubert and Stefan Schubert (2022), "SEER: Unsupervised and 
%    sample-efficient environment specialization of image descriptors", 
%    Proceeding of Robotics: Science And Systems, New York, USA
%
% This is a batch version that computes output descriptors for a batch of
% input descriptors Y.
% 
% M ... sparse exemplar memory, each COLUMN is an exemplar, can be empty. It
%       is important to allocate the appropriate amount of memory. If an
%       empty matrix is provided, then the memory is allocated within this
%       function.
% Y ... input descriptor, each row is a descriptor. Should be a distributed
%       representation.
% updatedM_flag ... toggels whether new rows/exemplars are added to M
% d_M ... number of non-zero elemente in the exemplars in M, default: 200
% k ... minimum number of exemplars per database descriptor, default: 50
% lambda ... factor on k for non-zero elements  in the output descriptor,
%            default: 2
% 
% M ... see input, will be updated if updatedM_flag is set
% DS ... sparse output descriptor
%
% peer.neubert@etit.tu-chemnitz.de, 2022
function [M, DS] = runSEER(M, Y, updateM_flag, d_M, k, lambda)

    % default parameters
    if ~exist('d_M','var') || isempty(d_M), d_M = 200; end
    if ~exist('k', 'var') || isempty(k), k = 50; end
    if ~exist('lambda', 'var') || isempty(lambda), lambda = 2; end    
    
    
    %% allocate memory
    % allocate memory for output descriptor
    DS = sparse([],[],[],size(Y,1), size(M,2), size(Y,1)*k);
    
    % Allocate memory for M if it is empty AND there has not already been 
    % allocated some memory 
    if isempty(M)
        propM = whos('M');
        if propM.bytes<=24 % 24 bytes is the minimum size of an empty sparse double matrix 
            M = sparse([],[],[],size(Y,2), 0, size(Y,1)*k*d_M); 
        end
    end
        
    %% L2 normalize each input descritptor
    Y = normr(Y);
    
    %% process each descriptor
    for i=1:size(Y,1)
        if mod(i, 100)==0; fprintf('  running image %d of %d (%d exemplars) \n', i, size(Y,1), size(M,2)); end
        
        t = tic;
        
        % compute similarity
        if ~isempty(M)            
            S = Y(i,:)*M; % dot product similarity                   
        else
            S = [];
        end
        
        % sparsify
        [knnSim, knnIdx] = maxk(S, lambda*k);


        if updateM_flag

            % apply threshold
            expectedSimilarity =  d_M / size(Y,2);    % activity=activity* size(Y,2)/nDimSamples;
            threshIdx = knnSim >= expectedSimilarity;
            knnIdxAboveThresh = knnIdx(threshIdx);
        
            % potentially create new patterns in M        
            if numel(knnIdxAboveThresh)< k
            
                nNewPatterns = k - numel(knnIdxAboveThresh); % the number of new exemplars           
                [PI, PV] = createExemplars(Y(i,:),d_M, nNewPatterns);
                newPatternIdx = size(M,2) + [1:size(PI,1)]; % indices of the new exemplars in M
                newPatternActivity = sum(PV.*PV, 2)'; % activity of the new exemplars
    
                for pIdx = 1:size(PI,1)
                    M(PI(pIdx,:), end+1) = PV(pIdx,:);
                end              
            else
                newPatternIdx = [];
                newPatternActivity = [];
                nNewPatterns = 0;
            end
        else
            newPatternIdx = [];
            newPatternActivity = [];
            nNewPatterns = 0;
        end

        % output descriptor is kNN results above thresh and new patterns
        nRemainingKNN = min(numel(knnIdx), (k*lambda)-nNewPatterns);
        descrIdx = [knnIdx(1:nRemainingKNN) newPatternIdx];
        descrActivity = [knnSim(1:nRemainingKNN) newPatternActivity];
                
        % create output descriptor
        DS(i, descrIdx) = descrActivity; 
       
    end
    

end

% Sample new exemplars
%
% inputY ... continuous input descriptor
%
% PI ... index matrix, each row is vector of size nDimSamples
% PV ... the values from inputY that correspond to these indexes 
%
function [PI, PV] = createExemplars(inputY,nDimSamples, nPatterns)

    PI = zeros(nPatterns, nDimSamples);
    for i=1:nPatterns
        
        if nDimSamples == size(inputY,2)
            PI(i,:) = 1:size(inputY,2); % use all dimensions
        else
            PI(i,:) = datasample(1:size(inputY,2), nDimSamples, 'Weights',abs(inputY),'Replace', false);
        end
        
    end
    PV = inputY(PI);

end